### PR TITLE
virt: use string table to detect VM or container

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

File: src/basic/virt.c [hunk 1]
```diff
@@ -521,29 +519,25 @@ int detect_container(void) {
...
-                        goto finish;
-                }
-
-        r = VIRTUALIZATION_CONTAINER_OTHER;
+        v = container_from_string(e);
+        if (v < 0)
+                v = VIRTUALIZATION_CONTAINER_OTHER;
 
 finish:
         log_debug("Found container virtualization %s.", virtualization_to_string(r));
```
Comment (ronnychevalier) [inline src/basic/virt.c:537]: I think you meant `virtualization_to_string(v)` here (`v` instead of `r`)?

Comment (yuwata) [reply src/basic/virt.c:537]: Ouch.

Comment (yuwata) [reply src/basic/virt.c:537]: Thank you for finding that. I've force-pushed the revised commit.

File: src/basic/virt.h [hunk 2]
```diff
@@ -5,7 +5,7 @@
 
 #include "macro.h"
 
-enum {
+typedef enum Virtualization {
```
Comment (poettering) [inline src/basic/virt.h]: btw, the reason this was an anonymous so far was that the main function returning this (`detect_virtualization()`) was supposed to also return negative error codes, and hence it made more sense using `int` for it, and don't bother with a name of the type. Not sure myself though whether that reasoning is strictly true though

File: src/basic/virt.c [hunk 3]
```diff
@@ -20,29 +20,28 @@
 #include "string-util.h"
 #include "virt.h"
 
+static const char *const vm_table[] = {
```
Comment (poettering) [inline src/basic/virt.c]: given this table is spare it really should be sized to `_VIRTUALIZATION_MAX` I think

File: src/basic/virt.c [hunk 4]
```diff
@@ -432,25 +431,24 @@ int detect_vm(void) {
...
-        } value_table[] = {
-                { "lxc",            VIRTUALIZATION_LXC            },
-                { "lxc-libvirt",    VIRTUALIZATION_LXC_LIBVIRT    },
-                { "systemd-nspawn", VIRTUALIZATION_SYSTEMD_NSPAWN },
-                { "docker",         VIRTUALIZATION_DOCKER         },
-                { "podman",         VIRTUALIZATION_PODMAN         },
-                { "rkt",            VIRTUALIZATION_RKT            },
-                { "wsl",            VIRTUALIZATION_WSL            },
-        };
+static const char *const container_table[] = {
```
Comment (poettering) [inline src/basic/virt.c]: as above

File: src/basic/virt.h (hunk 2 above)
Comment (yuwata) [reply src/basic/virt.h]: OK. This change is dropped.

**Comment by yuwata** [comment]
Updated. PTAL.

**Pull request merged**